### PR TITLE
EKF2: Terrain state

### PR DESCRIPTION
**Pull request by bresch**
### Solved Problem
- The terrain estimator depends on the EKF2 states so it's not possible to use a distance sensor or an optical flow in both estimators at the same time.
- Uncertainties of the altitude above ground are also not handled correctly when used for optical flow fusion.
- Range height fusion requires to estimate a height bias while this bias is the actual terrain height.

### Solution
Add a "terrain height" state to the EKF and modify the range and optical flow measurement jacobians to use it. This allows to simultaneously use those sensors for altitude, terrain height and velocity estimation. 
Using optical flow without a range sensor is also possible as the EKF would run as a poor man's VIO system.

The height above ground uncertainty is also now correctly handled and oscillations at higher altitudes due to over corrections from optical flow should be reduced. (To be tested)

### Changelog Entry
For release notes:
```
Integrate terrain estimator into EKF2
New parameter: remove EKF2_TERR_MASK
Documentation: -
```

### Test coverage
- Unit tests
- Indoor bench tests

   - Range + flow fusion:
Screenshot from 2024-05-29 17-35-06

   - Flow only (dist sensor data logged for comparison):
Screenshot from 2024-05-29 17-33-14

Indoor flight tests (range + flow and flow only)

TODO:
- [x] outdoor flight testing
- [x] names cleanup (rng_hgt vs hagl)
- [x] CONFIG_EKF2_TERRAIN cleanup

### CODE REVIEW

**Comment by bresch** [comment]
Flight tests:

with range and flow
Screenshot from 2024-06-20 10-57-34

same flight replayed with flow only (`EKF2_RNG_CTRL` disabled)
Screenshot from 2024-06-20 10-57-10

**Comment by dirksavage88** [comment]
If I understand the problem, the terrain height, velocity estimation, and altitude are all impacted if terrain hold is set under EKF2_RNG_CTRL? And this PR allows all three to use the terrain height state instead?

So if MPC_ALT_MODE is set to altitude following, this PR doesn't have an impact, only when terrain hold and optical flow (with range source) are enabled?

**Comment by AlexKlimaj** [comment]
Overall it flew really well. 

With GPS Fusion
https://review.px4.io/plot_app?log=329e00ef-098a-4336-b356-e2490fa80132

Without GPS Fusion
https://review.px4.io/plot_app?log=506aacf3-fe54-4c0f-96ac-b0ede8a5d7db

Without GPS Fusion. Over 40m above ground.
https://review.px4.io/plot_app?log=2f88cf4b-4daf-46ea-84c7-7694f842dff9

**Comment by bresch** [comment]
Other flights with large terrain height changes:
The dist bottom matches the range data pretty well and the velocity estimate is close to the GNSS one (GNSS is only there for logging, not fused in the EKF).
Screenshot from 2024-06-25 11-31-32
Screenshot from 2024-06-25 11-30-34

**Comment by bresch** [comment]
@dirksavage88 
> If I understand the problem, the terrain height, velocity estimation, and altitude are all impacted if terrain hold is set, EKF2_RNG_CTRL enabled, AND optical flow is fused? And this PR allows all three to use the terrain height state instead?

Terrain height, velocity and altitude are estimated when range and optical flow are fused, yes. Currently (in `main`), the terrain is estimated separately and EKF2 uses the terrain estimate to fuse optical flow. When fusing optical flow, EKF2 is not aware of the height above ground (hagl) uncertainty and can "over-fit" the measurements. By incorporating the terrain estimate in EKF2 those uncertainties are correctly handled and we can also avoid some circular dependencies (i.e.: when range data was used in EKF2 and the terrain estimator).

> So if MPC_ALT_MODE is set to altitude following, this PR doesn't have an impact, only when terrain hold and optical flow (with range source) are enabled?

From the outside, the behavior shouldn't really change, the estimates should just be "better". This, regardless of how `MPC_ALT_MODE` is set. And yes, there are no changes at all if no optical flow nor range finder sensor are used.

**Review by AlexKlimaj — DISMISSED** [review]

**Review by sfuhrer — APPROVED** [review]

File: src/modules/ekf2/EKF/covariance.cpp [hunk 1]
```diff
@@ -99,6 +99,11 @@ void Ekf::initialiseCovariance()
 #if defined(CONFIG_EKF2_WIND)
 	resetWindCov();
 #endif // CONFIG_EKF2_WIND
+
+#if defined(CONFIG_EKF2_TERRAIN)
+	// use the ground clearance value as our uncertainty
+	P.uncorrelateCovarianceSetVariance<State::terrain.dof>(State::terrain.idx, sq(_params.rng_gnd_clearance));
```
Comment (dagar) [inline src/modules/ekf2/EKF/covariance.cpp:105]: rng_gnd_clearance could be zero?

**Comment by AlexKlimaj** [comment]
Here are a couple more flights with GPS fusion enabled. Flight is looking really good.

https://review.px4.io/plot_app?log=4771723d-779d-47f4-bf61-351b8a7ca6bb
https://review.px4.io/plot_app?log=80260811-cdce-40c3-a2ba-3916526964d0

**Pull request merged**